src/pk.c

```diff
@@ -6959,793 +7110,1199 @@ WOLFSSL_BIGNUM* wolfSSL_DH_8192_prime(WOLFSSL_BIGNUM* bn)
 {
-    int ret;
-    word32 derSz = 0, pemSz = 0;
-    byte *der, *pem;
-    DhKey* key;
+    int ret = 1;
+    int errors = 0;
 
-    WOLFSSL_ENTER("wolfSSL_PEM_write_DHparams");
+    WOLFSSL_ENTER("wolfSSL_DH_check");
 
+    /* Validate parameters. */
     if (dh == NULL) {
-        WOLFSSL_LEAVE("wolfSSL_PEM_write_DHparams", BAD_FUNC_ARG);
-        return 0;
-    }
-
-    if (dh->inSet == 0) {
-        if (SetDhInternal(dh) != 1) {
-            WOLFSSL_MSG("Unable to set internal DH structure");
-            return 0;
-        }
-    }
-    key = (DhKey*)dh->internal;
-    ret = wc_DhParamsToDer(key, NULL, &derSz);
-    if (ret != LENGTH_ONLY_E) {
-        WOLFSSL_MSG("Failed to get size of DH params");
-        WOLFSSL_LEAVE("wolfSSL_PEM_write_DHparams", ret);
-        return 0;
+        ret = 0;
     }
 
-    der = (byte*)XMALLOC(derSz, key->heap, DYNAMIC_TYPE_TMP_BUFFER);
-    if (der == NULL) {
-        WOLFSSL_LEAVE("wolfSSL_PEM_write_DHparams", MEMORY_E);
-        return 0;
-    }
-    ret = wc_DhParamsToDer(key, der, &derSz);
-    if (ret <= 0) {
-        WOLFSSL_MSG("Failed to export DH params");
-        WOLFSSL_LEAVE("wolfSSL_PEM_write_DHparams", ret);
-        XFREE(der, key->heap, DYNAMIC_TYPE_TMP_BUFFER);
-        return 0;
+    /* Check generator available. */
+    if ((ret == 1) && ((dh->g == NULL) || (dh->g->internal == NULL))) {
+        errors |= DH_NOT_SUITABLE_GENERATOR;
     }
 
-    /* convert to PEM */
-    ret = wc_DerToPem(der, derSz, NULL, 0, DH_PARAM_TYPE);
-    if (ret < 0) {
-        WOLFSSL_MSG("Failed to convert DH params to PEM");
-        WOLFSSL_LEAVE("wolfSSL_PEM_write_DHparams", ret);
-        XFREE(der, key->heap, DYNAMIC_TYPE_TMP_BUFFER);
-        return ret;
+    if (ret == 1) {
+        /* Check prime available. */
+        if ((dh->p == NULL) || (dh->p->internal == NULL)) {
+            errors |= DH_CHECK_P_NOT_PRIME;
+        }
+        else {
+            /* Test if dh->p is prime. */
+            int isPrime = MP_NO;
+            ret = wolfssl_dh_check_prime(dh->p, &isPrime);
+            /* Set error if parameter p is not prime. */
+            if ((ret == 1) && (isPrime != MP_YES)) {
+                errors |= DH_CHECK_P_NOT_PRIME;
+            }
+        }
     }
-    pemSz = (word32)ret;
 
-    pem = (byte*)XMALLOC(pemSz, key->heap, DYNAMIC_TYPE_TMP_BUFFER);
-    if (pem == NULL) {
-        WOLFSSL_LEAVE("wolfSSL_PEM_write_DHparams", MEMORY_E);
-        XFREE(der, key->heap, DYNAMIC_TYPE_TMP_BUFFER);
-        return ret;
+    /* Return errors when user wants exact issues. */
+    if (codes != NULL) {
+        *codes = errors;
     }
-    ret = wc_DerToPem(der, derSz, pem, pemSz, DH_PARAM_TYPE);
-    XFREE(der, key->heap, DYNAMIC_TYPE_TMP_BUFFER);
-    if (ret < 0) {
-        WOLFSSL_MSG("Failed to convert DH params to PEM");
-        WOLFSSL_LEAVE("wolfSSL_PEM_write_DHparams", ret);
-        XFREE(pem, key->heap, DYNAMIC_TYPE_TMP_BUFFER);
-        return ret;
+    if (errors) {
+       ret = 0;
```

Comment:
Isn't this at odds with "OpenSSL does not return 0 when no operational failures buf error codes set."?

Reply:
Inconsistencies like this are annoying! (What are customers expecting?)
Changed to return error only when codes is NULL. OpenSSL assumes not NULL.
Updated tests.